include/fiction/io/read_fgl_layout.hpp

```diff
@@ -0,0 +1,813 @@
+            if (clocking_scheme_name != nullptr && clocking_scheme_name->GetText())
+            {
+                const auto clocking_scheme = get_clocking_scheme<Lyt>(clocking_scheme_name->GetText());
+                if (clocking_scheme.has_value())
+                {
+                    lyt.replace_clocking_scheme(*clocking_scheme);
+                    static constexpr std::array<const char*, 3> open_clocking_schemes{"OPEN", "OPEN3", "OPEN4"};
+
+                    auto* const clock_zones = clocking->FirstChildElement("zones");
+                    if (clock_zones != nullptr)
+                    {
+                        for (const auto* clock_zone = clock_zones->FirstChildElement("zone"); clock_zone != nullptr;
+                             clock_zone             = clock_zone->NextSiblingElement("zone"))
+                        {
+                            auto* const clocking_zone_x = clock_zone->FirstChildElement("x");
+                            int         x_coord         = 0;
+                            if (clocking_zone_x != nullptr && clocking_zone_x->GetText())
+                            {
+                                x_coord = std::stoi(clocking_zone_x->GetText());
+                            }
+                            else
+                            {
+                                throw fgl_parsing_error("Error parsing FGL file: no element 'x' in 'zone'");
+                            }
+
+                            auto* const clocking_zone_y = clock_zone->FirstChildElement("y");
+                            int         y_coord         = 0;
+                            if (clocking_zone_y != nullptr && clocking_zone_y->GetText())
+                            {
+                                y_coord = std::stoi(clocking_zone_y->GetText());
+                            }
+                            else
+                            {
+                                throw fgl_parsing_error("Error parsing FGL file: no element 'y' in 'zone'");
+                            }
+
+                            auto* const clocking_zone_clock = clock_zone->FirstChildElement("clock");
+                            uint8_t     clock               = 0;
+                            if (clocking_zone_clock != nullptr && clocking_zone_clock->GetText())
+                            {
+                                clock = static_cast<uint8_t>(*clocking_zone_clock->GetText());
+                            }
+                            else
+                            {
+                                throw fgl_parsing_error("Error parsing FGL file: no element 'clock' in 'zone'");
+                            }
+
+                            lyt.assign_clock_number({x_coord, y_coord}, clock);
+                        }
+                    }
+                    else if (std::find(std::begin(open_clocking_schemes), std::end(open_clocking_schemes),
```

Comment:
Is there a reason for using `std::begin` and `std::end` here over `.begin()` and `.end()` like in the other cases?

Also, when you use `std::find`, you may use `cbegin` and `cend`.